contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,557 @@
+   containing any error Kubelet might have encountered
+3. The device plugin start it's gRPC server if it did not recieve an
+   error
+
+Note that in the register request, the device plugin lists the labels
+it wants the node to advertise.
+If the labels are conflicting with pre-existing labels, Kubelet does
+not answer with a successfull `RegisterRequest`.
+
+In the current alpha version, labels will be used as properties for
+the device.
+
+## Unix Socket
+
+Device Plugins are expected to communicate with Kubelet through gRPC
+on an Unix socket.
+When starting the gRPC server, they are expected to create a unix socket
+at the following host path: `/var/run/kubernetes`.
+
+For non bare metal device plugin this means they will have to mount the folder
+as a volume in their pod spec ([see Installation](##installation)).
+
+Device plugins can expect to find the socket to register themselves on
+the host at the following path:
+`/var/run/Kubernetes/kubelet.sock`.
+
+## Protocol Overview
+
+When first registering themselves against Kubelet, the device plugin
+will send:
+  * The name of their unix socket
+  * [The API version against which they were built](#versioning).
+  * Their `ResourceName` they want to advertise
+
+Kubelet answers with the minimum version it supports and whether or
+not there was an error. The errors may include (but not limited to):
+  * API version not supported
+  * A device plugin already registered this `ResourceName`
+
+Kubelet will then interact with the plugin through the following functions:
+  * ListAndWatch: The device plugin advertises a list of Devices to Kubelet
+    and sends it again if the state of a Device changes
+  * `Allocate`: Called when creating a container with a list of devices
+
+![Process](device-plugin.png)
+
+
+## API Specification
+
+```go
+// Registration is the service advertised by the Kubelet
+// Only when Kubelet answers with a success code to a Register Request
+// may Device Plugins start their service
+service Registration {
+	rpc Register(RegisterRequest) returns (RegisterResponse) {}
+}
+
+// DevicePlugin is the service advertised by Device Plugins
+service DevicePlugin {
+	// ListWatch returns a stream of List of Devices
+	// Whenever a Device state change or a Device disapears, ListWatch
+	// returns the new list
+	rpc ListWatch(Empty) returns (stream ListWatchResponse) {}
+
+	// Allocate is called during container creation so that the Device
+	// Plugin can run device specific operations and instruct Kubelet
+	// of the steps to make the Device available in the container
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+
+message RegisterRequest {
+	// Version of the API the Device Plugin was built against
+	string version = 1;
+	// Name of the unix socket the device plugin is listening on
+	string endpoint = 2;
+	// Name of the Device advertised same as the one in the Device struct
+	string resourceName = 3;
+	// The labels to advertise at the node level
```

Comment:
Removing it.

>  If we don't have any scheduler to use these labels in any near term, I don't think we want to add them now.

Well the user could use labels instead of the hacky script that has to be launched before kubelet and advertised on the official doc with nvidia-smi:
https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/

But I guess this will be done later.